src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,164 @@
+{
+    [ExportWorkspaceService(typeof(IUnusedReferencesService), ServiceLayer.Default), Shared]
+    internal class UnusedReferencesService : IUnusedReferencesService
+    {
+        private static readonly ReferenceType[] _processingOrder = new[]
+        {
+            ReferenceType.Project,
+            ReferenceType.Package,
+            ReferenceType.Assembly
+        };
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public UnusedReferencesService()
+        {
+        }
+
+        public async Task<ImmutableArray<ReferenceInfo>> GetUnusedReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceInfo> references,
+            CancellationToken cancellationToken)
+        {
+            // Create a lookup of used assembly paths
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            if (compilation is null)
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            var usedAssemblyLookup = compilation.GetUsedAssemblyReferences(cancellationToken)
+                .Select(reference => reference.Display)
+                .OfType<string>()
+                .ToImmutableHashSet();
+
+            return GetUnusedReferences(usedAssemblyLookup, references);
+        }
+
+        internal static ImmutableArray<ReferenceInfo> GetUnusedReferences(
+            ImmutableHashSet<string> usedAssemblyLookup,
+            ImmutableArray<ReferenceInfo> references)
+        {
+            var unusedReferences = ImmutableArray.CreateBuilder<ReferenceInfo>();
+            var referencesByType = references.GroupBy(reference => reference.ReferenceType)
+                .ToImmutableDictionary(group => group.Key);
+
+            var remainingUsedAssemblyLookup = usedAssemblyLookup;
+
+            // We process the references in order by their type. This means we favor transitive references
+            // over direct references where possible.
+            foreach (var referenceType in _processingOrder)
+            {
+                if (!referencesByType.ContainsKey(referenceType))
+                {
+                    continue;
+                }
+
+                remainingUsedAssemblyLookup = AddUnusedReferences(
+                    referencesByType[referenceType].ToImmutableArray(),
+                    remainingUsedAssemblyLookup,
+                    unusedReferences);
+            }
+
+            return unusedReferences.ToImmutableArray();
+        }
+
+        private static ImmutableHashSet<string> AddUnusedReferences(
+            ImmutableArray<ReferenceInfo> references,
+            ImmutableHashSet<string> usedAssemblyLookup,
+            ImmutableArray<ReferenceInfo>.Builder unusedReferences)
+        {
+            // Determine which toplevel references have compilation assemblies in set of used assemblies.
```

Comment:
Seems to be complicated filtering/mapping going on here. If you can provide an example setup in the comments, it might make the logic more easy to understand.

Reply:
Added more detail with examples.